src/common/Throttle.h

```diff
@@ -330,4 +331,81 @@ class OrderedThrottle {
   uint32_t waiters = 0;
 };
 
+
+class TokenBucketThrottle {
+
+  struct Bucket {
+    CephContext *cct;
+    const std::string name;
+    std::atomic<uint64_t> remain = { 0 }, max = { 0 };
+
+    Bucket(CephContext *cct, const std::string& n, uint64_t m)
+      : cct(cct), name(n),
+	remain(m), max(m)
+    {
+    }
+
+    uint64_t get(uint64_t c);
+    uint64_t put(uint64_t c);
+    void set_max(uint64_t m);
+  };
+
+  struct Blocker {
+    uint64_t tokens_requested;
+    Context *ctx;
+
+    Blocker(uint64_t _tokens_requested, Context* _ctx)
+      : tokens_requested(_tokens_requested), ctx(_ctx) {}
+  };
+
+  CephContext *m_cct;
+  Bucket m_throttle;
+  uint64_t m_avg = 0;
+  SafeTimer *m_timer;
+  Mutex *m_timer_lock;
+  FunctionContext *m_token_ctx = nullptr;
+  Mutex m_blockers_lock;
+  list<Blocker> m_blockers;
+  Mutex m_lock;
+
+public:
+TokenBucketThrottle(CephContext *cct, uint64_t capacity, uint64_t avg,
+		    SafeTimer *timer, Mutex *timer_lock);
+
+~TokenBucketThrottle();
+
+template <typename T, typename I, void(T::*MF)(int, I*)>
+bool get(uint64_t c, T *handler, I *item) {
+  if (0 == m_throttle.max)
+    return false;
+
+  bool waited = false;
+
+  uint64_t got = 0;
+  {
+    Mutex::Locker lock(m_lock);
+    got = m_throttle.get(c);
+  }
+  if (got < c) {
+    // Not enough tokens, add a blocker for it.
+    Context *ctx = new FunctionContext([this, handler, item](int r) {
+	(handler->*MF)(r, item);
+      });
+    Mutex::Locker blockers_lock(m_blockers_lock);
```

Comment:
It seems like ```m_lock``` and ```m_blockers_lock``` can be combined since you acquire both in the same methods so there really cannot be a lock dependency issue.

Reply:
Yes, there would not be a lock dependency issue if we combine the two locks. but..... I did not do that because I fellow the Single Responsibility Principle in lock design. It's better to protect only ONE resource by a lock. So I think we should keep that, m_lock for m_throttle and m_blockers_lock for m_blockers_list.

Reply:
I think that over-complicates the code in this case since it's not like it's a necessity nor performance issue that warrants fine-grained locking.

Reply:
okey updated.